### PR TITLE
add boolean flag to run repo update in the background or not in Syncer.SyncRepo

### PR DESCRIPTION
**Pull request by mollylogue**
Background behind this change: In working to test https://github.com/sourcegraph/sourcegraph/pull/28522 I realized that the asynchronous nature of `syncer.SyncRepo` made testing unnecessarily difficult. It would be easier for this function to have synchronous behavior and then if we need the update to happen in the background, add the goroutine at a higher level for ease of testing these pieces. So, I'm proposing removing the goroutine from this function entirely. Below is my reasoning for why we can safely do that. 

Currently, this goroutine is never being touched. The only place in the code that `SyncRepo` is being called is the `repo-lookup` endpoint in `repo-updater` [query to validate this](https://sourcegraph.com/search?q=context%3Aglobal+repo%3Agithub.com%2Fsourcegraph%2Fsourcegraph+SyncRepo+-file%3A_test.go%24+patternType%3Aliteral+case%3Ayes)

If we look where the `RepoLookup` function in the repo updater client (which corresponds to this endpoint) is called, we see there are [3 cases](https://sourcegraph.com/search?q=context%3Aglobal+repo%3Agithub.com%2Fsourcegraph%2Fsourcegraph+RepoLookup+-file%3A_test.go%24+patternType%3Aliteral+case%3Ayes)

Cases:
1) [repos.Add](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/frontend/backend/repos.go?L147) since this is for adding new repos, `repo` will always be nil and `SyncRepo` will exit before the goroutine.
2) [set_user_public_repos.go](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/frontend/graphqlbackend/set_user_public_repos.go?L106) but if you look on line 102, you see that if the repo already exists we return, so we again have a nil repo and exit before the goroutine in `SyncRepo`
3) [dependency_indexing_scheduler.go](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/cmd/worker/internal/codeintel/indexing/dependency_indexing_scheduler.go?L150) This is the only one I'm not sure about, since we could potentially be updating existing repos. However, I still think that in this case the functionality should be synchronous as opposed to asynchronous but I'd like to validate that with someone who knows this part of the code already. 

UPDATE: Adding in a boolean flag to do the sync (for non-nil repos) in the background or not and setting it to `true` so this shouldn't change any functionality. Then when I call this function in https://github.com/sourcegraph/sourcegraph/pull/28522 I can set that boolean to `false`

### CODE REVIEW

**Comment by tsenart** [comment]
@mollylogue: Seems like you're taking lots of ownership over this area so I want to make you aware of previous plans / ideas that may be relevant to you.

1. First, we wanted to eventually get sunset repo-updater, and today we're much closer to that that we were when I first wrote https://github.com/sourcegraph/sourcegraph/issues/16197 (e.g. most random background jobs were already migrated to the `worker` service).
2. Second, more directly related with this PR, there is no particularly good reason for us to run `Syncer.SyncRepo` in repo-updater https://github.com/sourcegraph/sourcegraph/issues/23676. That would however require us to first make the git update scheduler (`scheduler.go`) store its state in Postgres, so that we could update that state from the frontend too, once we successfully call `SyncRepo` there.

**Review by indradhanush — COMMENTED** [review]
@mollylogue This is awesome deep diving with a great outcome! 

I read through the description and the changes and they make sense to me. The benefits from this change are clear.

The only concern I have is that I wonder what performance hit will `RepoLookup` take in terms of slowing down (perhaps noticeably?) for mono repos, as now we will do a blocking call to `rg.GetRepo` inside `Syncer.syncRepo` [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@34d066c81b9485c5e89e05323e691e54cd52134c/-/blob/internal/repos/syncer.go?L363&subtree=true). 

Unfortunately I do realize that this is also a very vague statement and to avoid a potential regression, maybe we should guard this change behind a feature flag (turned off by default)? That way we can quickly test this on dotcom while keeping the behaviour unchanged for our customers' instances.

I'd wait for @ryanslade's thoughts on this (he's back tomorrow) in case there are some other edge cases that I am not aware of.

Regarding your last point about this:

> dependency_indexing_scheduler.go This is the only one I'm not sure about, since we could potentially be updating existing repos. However, I still think that in this case the functionality should be synchronous as opposed to asynchronous but I'd like to validate that with someone who knows this part of the code already.

I agree with your assessment. But also that maybe someone from @sourcegraph/code-intel can take a look at this and verify?

**Comment by indradhanush** [comment]
Sidenote: I found it easier to view the changes with whitespaces disabled: https://github.com/sourcegraph/sourcegraph/pull/28655/files?diff=unified&w=1

**Comment by mollylogue** [comment]
> @mollylogue This is awesome deep diving with a great outcome!
> 
> I read through the description and the changes and they make sense to me. The benefits from this change are clear.
> 
> The only concern I have is that I wonder what performance hit will `RepoLookup` take in terms of slowing down (perhaps noticeably?) for mono repos, as now we will do a blocking call to `rg.GetRepo` inside `Syncer.syncRepo` [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@34d066c81b9485c5e89e05323e691e54cd52134c/-/blob/internal/repos/syncer.go?L363&subtree=true).
> 
> Unfortunately I do realize that this is also a very vague statement and to avoid a potential regression, maybe we should guard this change behind a feature flag (turned off by default)? That way we can quickly test this on dotcom while keeping the behaviour unchanged for our customers' instances.
> 
> I'd wait for @ryanslade's thoughts on this (he's back tomorrow) in case there are some other edge cases that I am not aware of.
> 
> Regarding your last point about this:
> 
> > dependency_indexing_scheduler.go This is the only one I'm not sure about, since we could potentially be updating existing repos. However, I still think that in this case the functionality should be synchronous as opposed to asynchronous but I'd like to validate that with someone who knows this part of the code already.
> 
> I agree with your assessment. But also that maybe someone from @sourcegraph/code-intel can take a look at this and verify?

@indradhanush I like your idea of adding a flag for this. I can look into that.

**Review by ryanslade — APPROVED** [review]
I think this change looks good and by adding the flag it should be safe.

I would also suggest checking with code-intel around whether this could be made synchronous. 

We could also add a prometheus counter inside the goroutine to check whether it ever gets triggered. If not, we can safely remove that whole code path.

**Review by indradhanush — COMMENTED** [review]
The feature flag for `background` appears to be hardcoded at the moment. How would we test this without requiring a new deployment with code changes to reverse the flag?

I was thinking of using a site-config based feature flag. For example if you look at usages of `EnableGithubInternalRepoVisibility` [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@248c328512b22ad76e25b54225e0c77ff0271e80/-/blob/schema/schema.go?L571:2&subtree=true#tab=references), we can change the behaviour on the fly without requiring a new deployment. 

Does that sound like a better idea?

**Comment by ryanslade** [comment]
The way I understand it is that the flag only needs to be false in tests

**Comment by mollylogue** [comment]
> The feature flag for `background` appears to be hardcoded at the moment. How would we test this without requiring a new deployment with code changes to reverse the flag?
> 
> I was thinking of using a site-config based feature flag. For example if you look at usages of `EnableGithubInternalRepoVisibility` [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@248c328512b22ad76e25b54225e0c77ff0271e80/-/blob/schema/schema.go?L571:2&subtree=true#tab=references), we can change the behaviour on the fly without requiring a new deployment.
> 
> Does that sound like a better idea?

So I was considering that approach, but it seemed like we have no need to have this flag at the top level. For now I just want to be able to switch it to false for my particular use case. If we decide that we'd like to remove the asynchronous functionality entirely, then I'd suggest actually implementing the feature flag so we can turn it on/off.

**Comment by mollylogue** [comment]
> I think this change looks good and by adding the flag it should be safe.
> 
> I would also suggest checking with code-intel around whether this could be made synchronous.
> 
> We could also add a prometheus counter inside the goroutine to check whether it ever gets triggered. If not, we can safely remove that whole code path.

Awesome. How would you feel about doing the prometheus counter in a follow-up PR? I just want to err on the side of merging smaller pieces of code.

**Comment by mollylogue** [comment]
> The way I understand it is that the flag only needs to be false in tests

I'll be specifying `false` here: https://github.com/sourcegraph/sourcegraph/pull/28522/files#diff-68c15ffbc343c39c63f3a8f0ab6d32b62885ea45f4c4f3f0f63e5c70ebfbb21eR44 so it'd be synchronous in the worker I'm adding (though obviously the worker itself is not synchronous). That way I'd be testing the exact code path that's actually running.

**Comment by indradhanush** [comment]
> So I was considering that approach, but it seemed like we have no need to have this flag at the top level. For now I just want to be able to switch it to false for my particular use case. If we decide that we'd like to remove the asynchronous functionality entirely, then I'd suggest actually implementing the feature flag so we can turn it on/off.

My only concern is that in case there's a regression, you will then need to go through the cycle of reverting the change to make the call async again and wait for the builds to pass on main and the rollout. With a site-level feature flag we can turn it on and off quickly. Also it enables us to turn it on only when you think you'd be at your desk to see over the changes in case you'll need to revert back. With the current approach, you'll need to time the merge and then ensure that you're still around for whenever your code is deployed.

**Comment by mollylogue** [comment]
> > So I was considering that approach, but it seemed like we have no need to have this flag at the top level. For now I just want to be able to switch it to false for my particular use case. If we decide that we'd like to remove the asynchronous functionality entirely, then I'd suggest actually implementing the feature flag so we can turn it on/off.
> 
> My only concern is that in case there's a regression, you will then need to go through the cycle of reverting the change to make the call async again and wait for the builds to pass on main and the rollout. With a site-level feature flag we can turn it on and off quickly. Also it enables us to turn it on only when you think you'd be at your desk to see over the changes in case you'll need to revert back. With the current approach, you'll need to time the merge and then ensure that you're still around for whenever your code is deployed.

That's fair, but I don't think this is intended to be turned off/on sometimes vs not like that. The `background=true` behavior preserves the legacy functionality, and then there's currently only new code added w/ `background=false`, so less of a need for a feature flag. I agree that if we do end up wanting to switch it on globally eventually, we should add a flag since it will be messing w/ legacy behavior and we'd want to have the ability to switch it on/off quickly.

File: internal/repos/syncer.go [hunk 1]
```diff
@@ -250,7 +250,7 @@ func (d Diff) Len() int {
 // because we don't sync our "cloud_default" code hosts in the background
 // since there are too many repos. Instead we use an incremental approach where we check for
 // changes everytime a user browses a repo.
-func (s *Syncer) SyncRepo(ctx context.Context, name api.RepoName) (repo *types.Repo, err error) {
+func (s *Syncer) SyncRepo(ctx context.Context, name api.RepoName, background bool) (repo *types.Repo, err error) {
```
Comment (ryanslade) [inline internal/repos/syncer.go:253]: Worth adding a doc comment to explain what the background flag is for

**Review by ryanslade — APPROVED** [review]
LGTM

**Review by indradhanush — APPROVED** [review]
> I agree that if we do end up wanting to switch it on globally eventually, we should add a flag since it will be messing w/ legacy behavior and we'd want to have the ability to switch it on/off quickly.

That's right. This is what I was thinking as well. :ok_hand: 

LGTM.
